daemons/based/based_transaction.c

```diff
@@ -0,0 +1,318 @@
+ */
+static inline void
+free_transaction(gpointer data)
+{
+    g_queue_free_full((GQueue *) data, (GDestroyNotify) free_request_data);
+}
+
+/*!
+ * \internal
+ * \brief Get a CIB client's uncommitted transaction, if any
+ *
+ * \param[in] client  Client to look up
+ *
+ * \return The client's uncommitted transaction, if any
+ *
+ * \note This function owns the return value. The caller must not free it.
+ */
+static inline GQueue *
+get_transaction(const pcmk__client_t *client)
+{
+    if (transactions == NULL) {
+        return NULL;
+    }
+    return g_hash_table_lookup(transactions, client->id);
+}
+
+/*!
+ * \internal
+ * \brief Check whether a CIB client has an uncommitted transaction
+ *
+ * \param[in] client  Client to look up
+ *
+ * \return \p true if the client has an uncommitted transaction, otherwise
+ *         \p false
+ */
+bool
+based_has_transaction(const pcmk__client_t *client)
+{
+    CRM_ASSERT(client != NULL);
+    return (get_transaction(client) != NULL);
+}
+
+/*!
+ * \internal
+ * \brief Create a new transaction for a given CIB client
+ *
+ * \param[in] client  Client to initiate a transaction for
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_init_transaction(const pcmk__client_t *client)
+{
+    CRM_ASSERT(client != NULL);
+
+    // A client can have at most one transaction at a time
+    if (based_has_transaction(client)) {
+        return pcmk_rc_already;
+    }
+
+    if (transactions == NULL) {
+        transactions = pcmk__strkey_table(NULL, free_transaction);
+    }
+    g_hash_table_insert(transactions, client->id, g_queue_new());
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Add a new CIB request to an existing transaction
+ *
+ * \param[in] client      Client whose transaction to extend
+ * \param[in] request     CIB request
+ * \param[in] privileged  If \p true, the request has write privileges
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_extend_transaction(pcmk__client_t *client, xmlNodePtr request,
+                         bool privileged)
+{
+    struct request_data *data = NULL;
+    GQueue *transaction = NULL;
+    const char *op = NULL;
+
+    CRM_ASSERT((client != NULL) && (request != NULL));
+
+    // Calling this function without an existing transaction is a bug
+    transaction = get_transaction(client);
+    CRM_ASSERT(transaction != NULL);
+
+    /* A list of unsupported ops is easier to maintain than a list of supported
+     * ops. Some of these are checked only for sanity and should be impossible
+     * for a client to pass directly.
+     *
+     * CIB sync requests are non-atomic and inherently asynchronous: we have to
+     * wait for replies from other nodes via the mainloop. The changes in a
+     * transaction are applied to the live CIB only if the entire transaction
+     * can be applied. We can't achieve this guarantee for sync requests.
+     */
+    op = crm_element_value(request, F_CIB_OPERATION);
+    if (pcmk__str_any_of(op,
```

Comment:
Also need to make sure `F_CIB_HOST` is not set to some other node.

It turns out that this request validation introduces a headache. If a developer submits an unsupported request while building a transaction, they won't know it was rejected unless they check the return code. Since the transaction was discarded, additional requests will be processed immediately instead of being queued. The first sign of trouble may be when they try to commit the transaction. By that time, they've accidentally applied changes.

I think it's better to add a new `cib_call_options` value called `cib_extend_transaction` or `cib_transaction` or something like that. This also gives the developer greater control. The developer will explicitly declare their intent to add a new request to a transaction, and they should check the return code to ensure that it was added successfully. If the `cib_extend_transaction` call option is **not** set, then the new request is processed immediately, and no attempt is made to add it to the client's transaction.

It's questionable whether an extension error should cause the transaction to be discarded immediately. An extension error should occur only in the event of a developer error (unsupported request) or a memory error. It could be convenient for all future calls with `cib_extend_transaction` to fail since no transaction exists, and for the commit to fail accordingly. That way, it's impossible for a transaction to be applied with a missing request due to a developer error. However, it may be better to let the developer decide how to respond to an extension failure (either by discarding the transaction or in some other way).

Reply:
On second thought, making non-extending calls while a transaction is open is probably a bad practice, leading to poor readability:
```
init_transaction(cib)
// return code checks elided
call1(cib, cib_extend_transaction)
call2(cib)
call3(cib, cib_extend_transaction)
commit_transaction(cib, cib_none)
```

Reply:
Darn... extending the transaction won't return a value to the client unless they call it with `cib_sync_call`. Will have to add everything to the transaction and do the validation at commit time.

Reply:
> I think it's better to add a new cib_call_options value called cib_extend_transaction or cib_transaction or something like that.

Definitely.

We could potentially drop the init transaction call, and just use the first call with cib_transaction to initialize a new transaction. Not sure whether that's more convenient or dangerous.

> It's questionable whether an extension error should cause the transaction to be discarded immediately. 

I could see it either way. On the one hand, the developer might want to commit any successful changes to that point, or try something different. On the other hand, atomicity would suggest any failure is a transaction failure.

> On second thought, devs should be checking the return code of each call anyway.

Definitely.

> And making non-extending calls while a transaction is open is probably a bad practice, leading to poor readability:

I don't think so. It's not necessarily in the same place; there could be a lot of async stuff going on, with a transaction being built gradually while normal calls proceed. Unlikely, but seems reasonable to support it.

> Darn... extending the transaction won't return a value to the client unless they call it with cib_sync_call. Will have to add everything to the transaction and do the validation at commit time.

pcmk_any_flags_set(flag, cib_sync_call|cib_transition)?

Reply:
> We could potentially drop the init transaction call, and just use the first call with cib_transaction to initialize a new transaction. Not sure whether that's more convenient or dangerous.

Mostly I like the symmetry (with commit) and clarity that `init_transaction()` gives. It also provides a bit of safety in case a developer forgets to commit a transaction and they intend to start a new one. If they do so via an "extends" call option, then it'll just be added to their existing transaction. If they do so via the init function, it will return `pcmk_rc_already`. (An unpublished early approach did use call options for everything BTW.)

> I could see it either way. On the one hand, the developer might want to commit any successful changes to that point, or try something different. On the other hand, atomicity would suggest any failure is a transaction failure.

The current draft does the validation at commit time. This turns out to be simpler: we treat an unsupported request exactly like a failed request. This comes at the expense of processing supported requests when an unsupported one is coming later, but that's not the common case.

> Unlikely, but seems reasonable to support it.

Yeah, I agree. In general I don't want to impose arbitrary restrictions. Another reason I chose to use functions rather than flags for init/commit is that the functions naturally allow trivial commits containing zero or one request.

We have enough non-arbitrary restrictions (for example, no calls that involve other nodes directly or indirectly).

> pcmk_any_flags_set(flag, cib_sync_call|cib_transition)?

Ah-ha, I'm not the only one who keeps typing "transition" out of habit.

Handling unsupported request failures at commit time renders this moot. This might work but would need to be done at the client side and feels a tad messier.

Reply:
> Handling unsupported request failures at commit time renders this moot.

Okay, that's not true. I still have this for `ENOMEM`:
```
            /* This should happen only on memory error. We have no way to notify
             * the client that extension failed (unless cib_sync_call was
             * given). Discard so that an incompletely built transaction is not
             * committed.
             */
            crm_err("Could not extend transaction for client %s, discarding: "
                    "%s",
                    pcmk__client_name(cib_client), pcmk_rc_str(rc));
            based_discard_transaction(cib_client);
```

Anyway, IIRC another reason was that sending a response to the client happens in `cib_process_request()` or `cib_process_command()`. We don't process a request that has `cib_extends_transaction` set. So even sending the reply from server side would get more complex.

Reply:
I decided it's best to at least send the reply at "extend transaction" time. An upcoming push attempts to do this. Checking the return code via callbacks and reacting it is super clunky, but it's best to have that option. The client can always make synchronous calls.